frame/nft-fractionalisation/src/lib.rs

```diff
@@ -0,0 +1,294 @@
+			> + NonFungiblesTransfer<Self::AccountId>;
+
+		#[pallet::constant]
+		type PalletId: Get<PalletId>;
+
+		#[pallet::constant]
+		type BuybackThreshold: Get<u32>;
+	}
+
+	#[pallet::storage]
+	#[pallet::getter(fn assets_minted)]
+	// TODO: query amount minted from pallet assets instead of storing it locally.
+	// Add a public getter function to pallet assets.
+	pub type AssetsMinted<T: Config> =
+		StorageMap<_, Twox64Concat, AssetIdOf<T>, AssetBalanceOf<T>, OptionQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn asset_to_nft)]
+	// TODO: store information about Asset ID and the corresponding Collection and Item ID.
+	pub type AssetToNft<T: Config> =
+		StorageMap<_, Twox64Concat, AssetIdOf<T>, (T::CollectionId, T::ItemId), OptionQuery>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		NFTLocked(T::CollectionId, T::ItemId),
+		NFTUnlocked(T::CollectionId, T::ItemId),
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		AssetAlreadyRegistered,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::weight(10_000 + T::DbWeight::get().writes(2).ref_time())]
+		/// Pallet's account must be funded before lock is possible!
+		/// 5EYCAe5gjC5dxKPbV2GPQUetETjFNSYZsSwSurVTTXidSLbh
+		pub fn lock_nft_create_asset(
+			origin: OriginFor<T>,
+			collection_id: T::CollectionId,
+			item_id: T::ItemId,
+			asset_id: AssetIdOf<T>,
+			beneficiary: T::AccountId,
+			min_balance: AssetBalanceOf<T>,
+			amount: AssetBalanceOf<T>,
+		) -> DispatchResult {
+			let _who = ensure_signed(origin.clone())?;
+			let admin_account_id = Self::pallet_account_id();
+
+			Self::do_lock_nft(collection_id, item_id)?;
+			Self::do_create_asset(asset_id, admin_account_id, min_balance)?;
+			Self::do_mint_asset(asset_id, &beneficiary, amount)?;
+
+			// Mutate this storage item to retain information about the amount minted.
+			<AssetsMinted<T>>::try_mutate(
+				asset_id,
+				|assets_minted| -> Result<(), DispatchError> {
+					match assets_minted.is_some() {
+						true =>
+							*assets_minted = Some(assets_minted.unwrap().saturating_add(amount)),
+						false => *assets_minted = Some(amount),
+					}
+
+					Ok(())
+				},
+			)?;
+
+			// Mutate this storage item to retain information about the asset created.
+			<AssetToNft<T>>::try_mutate(asset_id, |nft_id| -> Result<(), DispatchError> {
+				*nft_id = Some((collection_id, item_id));
+
+				Ok(())
+			})?;
+
+			Self::deposit_event(Event::NFTLocked(collection_id, item_id));
+
+			Ok(())
+		}
+
+		/// Return and burn a % of the asset, unlock the NFT. Currently 100% is the minimum
+		/// threshold.
+		#[pallet::weight(10_000 + T::DbWeight::get().writes(2).ref_time())]
+		pub fn burn_asset_unlock_nft(
+			origin: OriginFor<T>,
+			asset_id: AssetIdOf<T>,
+			amount: AssetBalanceOf<T>,
+		) -> DispatchResult {
+			let who = ensure_signed(origin.clone())?;
+
+			let (collection_id, item_id) = Self::get_nft_id(asset_id);
```

Comment:
If i'm not wrong in case the method `get_nft_id` throws an error because of the unwrap, the error is not being handled. I would catch the error and throw a proper `Error`